From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Fri, 18 Mar 2022 03:30:44 +0100
Subject: background: Support changing the light / dark theme from g-c-c

Without this we'd not be able to change the background when the dark
theme has been chosen.
---
 panels/background/cc-background-chooser.c  | 18 ++++++++++++++++
 panels/background/cc-background-chooser.h  |  2 ++
 panels/background/cc-background-chooser.ui | 13 ++++++++++++
 panels/background/cc-background-panel.c    | 33 ++++++++++++++++++++++++++++--
 4 files changed, 64 insertions(+), 2 deletions(-)

diff --git a/panels/background/cc-background-chooser.c b/panels/background/cc-background-chooser.c
index 04fd85c..97aea0b 100644
--- a/panels/background/cc-background-chooser.c
+++ b/panels/background/cc-background-chooser.c
@@ -36,6 +36,7 @@ struct _CcBackgroundChooser
   GtkFlowBox         *flowbox;
   GtkWidget          *recent_box;
   GtkFlowBox         *recent_flowbox;
+  GtkLabel           *bg_notice;
 
   gboolean            recent_selected;
 
@@ -281,6 +282,22 @@ on_file_chooser_selection_changed_cb (GtkFileChooser               *chooser,
   gtk_file_chooser_set_preview_widget_active (chooser, TRUE);
 }
 
+void
+cc_background_chooser_set_dark_mode (CcBackgroundChooser *self,
+                                     gboolean             is_dark)
+{
+  if (is_dark)
+    {
+      gtk_label_set_label (self->bg_notice,
+                           _("This background selection only applies to the dark style"));
+    }
+  else
+    {
+      gtk_label_set_label (self->bg_notice,
+                           _("This background selection only applies to the light style"));
+    }
+}
+
 /* GObject overrides */
 
 static void
@@ -315,6 +332,7 @@ cc_background_chooser_class_init (CcBackgroundChooserClass *klass)
   gtk_widget_class_bind_template_child (widget_class, CcBackgroundChooser, flowbox);
   gtk_widget_class_bind_template_child (widget_class, CcBackgroundChooser, recent_box);
   gtk_widget_class_bind_template_child (widget_class, CcBackgroundChooser, recent_flowbox);
+  gtk_widget_class_bind_template_child (widget_class, CcBackgroundChooser, bg_notice);
 
   gtk_widget_class_bind_template_callback (widget_class, on_item_activated_cb);
 }
diff --git a/panels/background/cc-background-chooser.h b/panels/background/cc-background-chooser.h
index 1751098..043df64 100644
--- a/panels/background/cc-background-chooser.h
+++ b/panels/background/cc-background-chooser.h
@@ -28,5 +28,7 @@ G_BEGIN_DECLS
 G_DECLARE_FINAL_TYPE (CcBackgroundChooser, cc_background_chooser, CC, BACKGROUND_CHOOSER, GtkBox)
 
 void cc_background_chooser_select_file (CcBackgroundChooser *self);
+void cc_background_chooser_set_dark_mode (CcBackgroundChooser *self,
+                                          gboolean             is_dark);
 
 G_END_DECLS
diff --git a/panels/background/cc-background-chooser.ui b/panels/background/cc-background-chooser.ui
index aa7da28..422418e 100644
--- a/panels/background/cc-background-chooser.ui
+++ b/panels/background/cc-background-chooser.ui
@@ -5,6 +5,19 @@
     <property name="can_focus">False</property>
     <property name="orientation">vertical</property>
 
+    <child>
+      <object class="GtkLabel" id="bg_notice">
+        <property name="visible">True</property>
+        <property name="can_focus">False</property>
+        <property name="visible">1</property>
+        <property name="xalign">0.5</property>
+        <property name="margin-bottom">12</property>
+        <style>
+          <class name="dim-label"/>
+        </style>
+      </object>
+    </child>
+
     <child>
       <object class="GtkSeparator">
         <property name="visible">True</property>
diff --git a/panels/background/cc-background-panel.c b/panels/background/cc-background-panel.c
index 3e50f50..e6525b5 100644
--- a/panels/background/cc-background-panel.c
+++ b/panels/background/cc-background-panel.c
@@ -36,12 +36,16 @@
 #include "cc-background-xml.h"
 
 #define WP_PATH_ID "org.gnome.desktop.background"
+#define WP_INTERFACE_ID "org.gnome.desktop.interface"
 #define WP_LOCK_PATH_ID "org.gnome.desktop.screensaver"
 #define WP_URI_KEY "picture-uri"
+#define WP_URI_DARK_KEY "picture-uri-dark"
 #define WP_OPTIONS_KEY "picture-options"
 #define WP_SHADING_KEY "color-shading-type"
 #define WP_PCOLOR_KEY "primary-color"
 #define WP_SCOLOR_KEY "secondary-color"
+#define COLOR_SCHEME_KEY "color-scheme"
+
 
 struct _CcBackgroundPanel
 {
@@ -50,6 +54,7 @@ struct _CcBackgroundPanel
   GDBusConnection *connection;
 
   GSettings *settings;
+  GSettings *interface_settings;
   GSettings *lock_settings;
 
   GnomeDesktopThumbnailFactory *thumb_factory;
@@ -82,6 +87,12 @@ get_save_path (void)
                            NULL);
 }
 
+static gboolean
+in_dark_mode (CcBackgroundPanel *self)
+{
+  return g_str_equal(g_settings_get_string (self->interface_settings, COLOR_SCHEME_KEY), "prefer-dark");
+}
+
 static void
 reload_current_bg (CcBackgroundPanel *panel)
 {
@@ -98,7 +109,18 @@ reload_current_bg (CcBackgroundPanel *panel)
 
   /* initalise the current background information from settings */
   settings = panel->settings;
-  uri = g_settings_get_string (settings, WP_URI_KEY);
+
+  if (in_dark_mode (panel))
+    {
+      cc_background_chooser_set_dark_mode (panel->background_chooser, TRUE);
+      uri = g_settings_get_string (settings, WP_URI_DARK_KEY);
+    }
+  else
+    {
+      cc_background_chooser_set_dark_mode (panel->background_chooser, FALSE);
+      uri = g_settings_get_string (settings, WP_URI_KEY);
+    }
+
   if (uri && *uri == '\0')
     g_clear_pointer (&uri, g_free);
 
@@ -167,7 +189,10 @@ set_background (CcBackgroundPanel *panel,
   uri = cc_background_item_get_uri (item);
   flags = cc_background_item_get_flags (item);
 
-  g_settings_set_string (settings, WP_URI_KEY, uri);
+  if (settings == panel->settings && in_dark_mode (panel))
+    g_settings_set_string (settings, WP_URI_DARK_KEY, uri);
+  else
+    g_settings_set_string (settings, WP_URI_KEY, uri);
 
   /* Also set the placement if we have a URI and the previous value was none */
   if (flags & CC_BACKGROUND_ITEM_HAS_PLACEMENT)
@@ -238,6 +263,7 @@ cc_background_panel_dispose (GObject *object)
 
   g_clear_object (&panel->settings);
   g_clear_object (&panel->lock_settings);
+  g_clear_object (&panel->interface_settings);
   g_clear_object (&panel->thumb_factory);
 
   G_OBJECT_CLASS (cc_background_panel_parent_class)->dispose (object);
@@ -303,10 +329,13 @@ cc_background_panel_init (CcBackgroundPanel *panel)
   panel->lock_settings = g_settings_new (WP_LOCK_PATH_ID);
   g_settings_delay (panel->lock_settings);
 
+  panel->interface_settings = g_settings_new (WP_INTERFACE_ID);
+
   /* Load the background */
   reload_current_bg (panel);
   update_preview (panel);
 
   /* Background settings */
   g_signal_connect_object (panel->settings, "changed", G_CALLBACK (on_settings_changed), panel, G_CONNECT_SWAPPED);
+  g_signal_connect_object (panel->interface_settings, "changed::"COLOR_SCHEME_KEY, G_CALLBACK (on_settings_changed), panel, G_CONNECT_SWAPPED);
 }
